pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java

```diff
@@ -0,0 +1,174 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.requesthandler;
+
+import com.google.common.base.Preconditions;
+import java.util.Collections;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import org.apache.helix.HelixAdmin;
+import org.apache.helix.HelixConstants;
+import org.apache.helix.HelixManager;
+import org.apache.helix.model.HelixConfigScope;
+import org.apache.helix.model.builder.HelixConfigScopeBuilder;
+import org.apache.pinot.broker.broker.helix.ClusterChangeHandler;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class helps limit the number of multi-stage queries being executed concurrently. Currently, this limit is
+ * applied at the broker level, but could be moved to the server level in the future.
+ */
+public class MultiStageQuerySemaphore extends Semaphore implements ClusterChangeHandler {
```

Comment:
IMHO this class does too many things. I would suggest to split it in two different classes:

1. The semaphore itself. It doesn't need to know the number of brokers, max concurrent queries, etc. It has a `setTotalPermits` for a new number of permits, which calls `super.reducePermits` and `super.release` if needed.
2. The ClusterChangeHandler, which has a reference to the semaphore. It knows the number of brokers and max concurrent queries and calls `setTotalPermits` each time there are cluster changes.

By doing so the semaphore has no knowledge related to this feature and it is just a semaphore with the ability to change permits. It is easier to understand and we can use it in other developments. And different handlers can be used as needed.

Reply:
Probably you can even generalize the handler to depend on some function callbacks or abstract methods so you don't need to mock all the helix attributes. I would prefer that (the less mocking the better) but I understand it can also be seen as overengineering.